Duplicati/Library/Backend/Jottacloud/Jottacloud.cs

```diff
@@ -0,0 +1,405 @@
+            {
+                System.Net.HttpWebRequest req = CreateRequest(System.Net.WebRequestMethods.Http.Post, "", "mkDir=true", false);
+                Utility.AsyncHttpRequest areq = new Utility.AsyncHttpRequest(req);
+                using (System.Net.HttpWebResponse resp = (System.Net.HttpWebResponse)areq.GetResponse())
+                { }
+            }
+        }
+
+        #endregion
+
+        #region IDisposable Members
+
+        public void Dispose()
+        {
+        }
+
+        #endregion
+
+        private System.Net.HttpWebRequest CreateRequest(string method, string url, string queryparams)
+        {
+            System.Net.HttpWebRequest req = (System.Net.HttpWebRequest)System.Net.HttpWebRequest.Create(url + (string.IsNullOrEmpty(queryparams) || queryparams.Trim().Length == 0 ? "" : "?" + queryparams));
+            req.Method = method;
+            req.Credentials = m_userInfo;
+            //We need this under Mono for some reason,
+            // and it appears some servers require this as well
+            req.PreAuthenticate = true;
+
+            req.KeepAlive = false;
+            req.UserAgent = "Duplicati Jottacloud Client v" + System.Reflection.Assembly.GetExecutingAssembly().GetName().Version.ToString();
+            req.Headers.Add("X-JottaAPIVersion", API_VERSION);
+
+            return req;
+        }
+
+        private System.Net.HttpWebRequest CreateRequest(string method, string remotename, string queryparams, bool upload)
+        {
+            var url = (upload ? m_url_upload : m_url) + Library.Utility.Uri.UrlEncode(remotename).Replace("+", "%20");
+            return CreateRequest(method, url, queryparams);
+        }
+
+        #region IStreamingBackend Members
+
+        public bool SupportsStreaming
+        {
+            get { return true; }
+        }
+
+        public void Get(string remotename, System.IO.Stream stream)
+        {
+            var req = CreateRequest(System.Net.WebRequestMethods.Http.Get, remotename, "mode=bin", false);
+            var areq = new Utility.AsyncHttpRequest(req);
+            using (var resp = (System.Net.HttpWebResponse)areq.GetResponse())
+            using (var s = areq.GetResponseStream())
+                Utility.Utility.CopyStream(s, stream, true, m_copybuffer);
+        }
+
+        public void Put(string remotename, System.IO.Stream stream)
+        {
+            if (!stream.CanSeek)
+            {
+                throw new System.Net.WebException(Strings.Jottacloud.FileUploadError, System.Net.WebExceptionStatus.ProtocolError);
+            }
+
+            // Pre-calculate MD5 hash, we need it in query parameter, in HTTP header and in POST message data!
+            string md5Hash;
+            using (System.Security.Cryptography.MD5 md5 = System.Security.Cryptography.MD5.Create())
+                md5Hash = BitConverter.ToString(md5.ComputeHash(stream)).Replace("-", string.Empty);
```

Comment:
The stream you get here can be throttled, which will make the hash computation "throttled".
There is another backend that needs the same, and it "walks" the streams, until it gets the underlying FileStream and computes the hash on that, such that it can be done without the throttle.